src/callback.rs

```diff
@@ -115,50 +117,63 @@ wrapping_cast!(i32, Py_hash_t);
 wrapping_cast!(isize, Py_hash_t);
 wrapping_cast!(i64, Py_hash_t);
 
-pub struct HashConverter<T>(pub std::marker::PhantomData<T>);
+pub struct HashConverter<T>(pub T);
 
-impl<T> CallbackConverter for HashConverter<T>
+impl<T> IntoPyCallbackResult<Py_hash_t> for HashConverter<T>
 where
     T: WrappingCastTo<Py_hash_t>,
 {
-    type Source = T;
-    type Result = Py_hash_t;
-    const ERR_VALUE: Self::Result = -1;
-
     #[inline]
-    fn convert(val: T, _py: Python) -> Py_hash_t {
-        let hash = val.wrapping_cast();
+    fn convert(self, _py: Python) -> PyResult<Py_hash_t> {
+        let hash = self.0.wrapping_cast();
         if hash == -1 {
-            -2
+            Ok(-2)
         } else {
-            hash
+            Ok(hash)
         }
     }
 }
 
-// Short hands methods for macros
+#[doc(hidden)]
 #[inline]
-pub fn cb_convert<C, T>(_c: C, py: Python, value: PyResult<T>) -> C::Result
+pub fn convert<T, U>(py: Python, value: T) -> PyResult<U>
 where
-    C: CallbackConverter<Source = T>,
+    T: IntoPyCallbackResult<U>,
 {
-    C::convert_result(py, value)
-}
-
-// Same as cb_convert(PyObjectCallbackConverter<T>, py, value)
-#[inline]
-pub fn cb_obj_convert<T: IntoPy<PyObject>>(
-    py: Python,
-    value: PyResult<T>,
-) -> <PyObjectCallbackConverter<T> as CallbackConverter>::Result {
-    PyObjectCallbackConverter::<T>::convert_result(py, value)
+    value.convert(py)
 }
 
+#[doc(hidden)]
 #[inline]
-pub unsafe fn cb_err<C>(_c: C, py: Python, err: impl Into<crate::PyErr>) -> C::Result
+pub fn callback_error<T>() -> T
 where
-    C: CallbackConverter,
+    T: PyCallbackResult,
 {
-    err.into().restore(py);
-    C::ERR_VALUE
+    T::ERR_VALUE
+}
+
+#[macro_export]
+macro_rules! catch_unwind {
+    ($py:ident, $block:expr) => {{
+        let result = match std::panic::catch_unwind(|| -> PyResult<_> { $block }) {
+            Ok(result) => result,
+            Err(e) => {
+                // Try to format the error in the same way panic does
+                if let Some(string) = e.downcast_ref::<String>() {
+                    Err($crate::exceptions::RuntimeError::py_err((string.clone(),)))
```

Comment:
Looks good, but isn't it better if we have a special exception type?

Reply:
In theory I like a special exception type, but I am not sure it is helpful in practice.

If we make an exception `RustPanicException`, how do we `import` it in Python code to catch it? Without the ability to name the exception in Python, they will be forced to write things such as `catch Exception as e`, which is bad practice because they're then catching all exceptions and not just panics.

I used to use C++ / boost::Python, which defines custom exception types for its `ArgumentError`, which was really frustrating when you then cannot catch this exception.

How we solve the `import pyo3.RustPanicException` gets even more complicated as we consider the possibility that multiple python libraries might be using different `pyo3` versions, and each version might have its own version of `RustPanicException`.

If we can come up with a good solution for this `import` problem, I am in favour of making a custom exception type. Until then, I want to stick to `RuntimeError`.

Reply:
OK, it's convincing.